File: libaums/src/com/github/mjdev/libaums/fs/fat32/FatDirectory.java
package com.github.mjdev.libaums.fs.fat32;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

import android.util.Log;

import com.github.mjdev.libaums.driver.BlockDeviceDriver;
import com.github.mjdev.libaums.fs.UsbFile;

public class FatDirectory implements UsbFile {
	
	private static String TAG = FatDirectory.class.getSimpleName();

	private ClusterChain chain;
	private BlockDeviceDriver blockDevice;
	private FAT fat;
	private Fat32BootSector bootSector;
	private List<FatLfnDirectoryEntry> entries;
	private Map<String, FatLfnDirectoryEntry> lfnMap;
	private Map<ShortName, FatDirectoryEntry> shortNameMap;
	private FatDirectory parent;
	private FatLfnDirectoryEntry entry;
	
	private String volumeLabel;
	
	private FatDirectory(BlockDeviceDriver blockDevice, FAT fat,
			Fat32BootSector bootSector, FatDirectory parent) {
		this.blockDevice = blockDevice;
		this.fat = fat;
		this.bootSector = bootSector;
		this.parent = parent;
		entries = new ArrayList<FatLfnDirectoryEntry>();
		lfnMap = new HashMap<String, FatLfnDirectoryEntry>();
		shortNameMap = new HashMap<ShortName, FatDirectoryEntry>();
	}
	
	/* package */ static FatDirectory create(FatLfnDirectoryEntry entry, BlockDeviceDriver blockDevice, FAT fat, Fat32BootSector bootSector, FatDirectory parent) throws IOException {
		FatDirectory result = new FatDirectory(blockDevice, fat, bootSector, parent);
		result.entry = entry;
		return result;
	}

	/* package */ static FatDirectory readRoot(BlockDeviceDriver blockDevice, FAT fat, Fat32BootSector bootSector) throws IOException {
		FatDirectory result = new FatDirectory(blockDevice, fat, bootSector, null);
		result.chain = new ClusterChain(bootSector.getRootDirStartCluster(), blockDevice, fat, bootSector);
		result.init();
		return result;
	}
	
	private void init() throws IOException {
		if(chain == null) {
			chain = new ClusterChain(entry.getStartCluster(), blockDevice, fat, bootSector);
		}

		if(entries.size() == 0)
			readEntries();
	}
	
	private void readEntries() throws IOException {
		ByteBuffer buffer = ByteBuffer.allocate((int)chain.getLength());
		chain.read(0, buffer);
		ArrayList<FatDirectoryEntry> list = new ArrayList<FatDirectoryEntry>();
		buffer.flip();
		while(buffer.remaining() > 0) {
			FatDirectoryEntry e = FatDirectoryEntry.read(buffer);
			if(e == null) {
				break;
			}
			
			if(e.isLfnEntry()) {
				list.add(e);
				continue;
			}
			
			if(e.isVolumeLabel()) {
				if(!isRoot()) {
					Log.w(TAG, "volume label in non root dir!");
				}
				volumeLabel = e.getVolumeLabel();
				Log.d(TAG, "volume label: " + volumeLabel);
				continue;
			}
			
			if(e.isDeleted()) {
				list.clear();
				continue;
			}
			
			FatLfnDirectoryEntry lfnEntry = FatLfnDirectoryEntry.read(e, list);
			addEntry(lfnEntry, e);
			list.clear();
		}
	}
	
	private void addEntry(FatLfnDirectoryEntry lfnEntry, FatDirectoryEntry entry) {
		entries.add(lfnEntry);
		lfnMap.put(lfnEntry.getName().toLowerCase(Locale.getDefault()), lfnEntry);
		shortNameMap.put(entry.getShortName(), entry);
	}
	
	/* package */  void removeEntry(FatLfnDirectoryEntry lfnEntry) {
		entries.remove(lfnEntry);
		lfnMap.remove(lfnEntry.getName().toLowerCase(Locale.getDefault()));
		shortNameMap.remove(lfnEntry.getActualEntry().getShortName());
	}
	
	/* package */ void renameEntry(FatLfnDirectoryEntry lfnEntry, String newName) throws IOException {
		if(lfnEntry.getName().equals(newName)) return;
		
		removeEntry(lfnEntry);
		lfnEntry.setName(newName, ShortNameGenerator.generateShortName(newName, shortNameMap.keySet()));
		addEntry(lfnEntry, lfnEntry.getActualEntry());
		write();
	}
	
	/* package */ void write() throws IOException {
		init();
		final boolean writeVolumeLabel = isRoot() && volumeLabel != null;
		// first lookup the total entries needed
		int totalEntryCount = 0;
		for(FatLfnDirectoryEntry entry : entries) {
			totalEntryCount += entry.getEntryCount();
		}
		
		if(writeVolumeLabel)
			totalEntryCount++;
		
		long totalBytes = totalEntryCount * FatDirectoryEntry.SIZE;
		chain.setLength(totalBytes);
		
		ByteBuffer buffer = ByteBuffer.allocate((int)chain.getLength());
		buffer.order(ByteOrder.LITTLE_ENDIAN);
		
		if(writeVolumeLabel)
			FatDirectoryEntry.createVolumeLabel(volumeLabel).serialize(buffer);
		
		for(FatLfnDirectoryEntry entry : entries) {
			entry.serialize(buffer);
		}
		
		if(totalBytes % bootSector.getBytesPerCluster() != 0) {
			// add dummy entry filled with zeros to mark end of entries
			buffer.put(new byte[32]);
		}
		
		buffer.flip();
		chain.write(0, buffer);
	}
	
	private boolean isRoot() {
		return entry == null;
	}
	
	/* package */ String getVolumeLabel() {
		return volumeLabel;
	}

	@Override
	public FatFile createFile(String name) throws IOException {
		if(lfnMap.containsKey(name.toLowerCase(Locale.getDefault()))) throw new IOException("Item already exists!");
		
		ShortName shortName = ShortNameGenerator.generateShortName(name, shortNameMap.keySet());
		
		FatLfnDirectoryEntry entry = FatLfnDirectoryEntry.createNew(name, shortName);
		long newStartCluster = fat.alloc(new Long[0], 1)[0];
		entry.setStartCluster(newStartCluster);
		
		Log.d(TAG, "adding entry: " + entry + " with short name: " + shortName);
		addEntry(entry, entry.getActualEntry());
		write();
		
		return FatFile.create(entry, blockDevice, fat, bootSector, this);
	}
	
	@Override
	public FatDirectory createDirectory(String name) throws IOException {
		if(lfnMap.containsKey(name.toLowerCase(Locale.getDefault()))) throw new IOException("Item already exists!");

		ShortName shortName = ShortNameGenerator.generateShortName(name, shortNameMap.keySet());
		
		FatLfnDirectoryEntry entry = FatLfnDirectoryEntry.createNew(name, shortName);
		entry.setDirectory();
		long newStartCluster = fat.alloc(new Long[0], 1)[0];
		entry.setStartCluster(newStartCluster);
		
		Log.d(TAG, "adding entry: " + entry + " with short name: " + shortName);
		addEntry(entry, entry.getActualEntry());
		write();
		
		FatDirectory result = FatDirectory.create(entry, blockDevice, fat, bootSector, this);
		
		// first create the dot entry which points to the dir just created
		FatLfnDirectoryEntry dotEntry = FatLfnDirectoryEntry.createNew(null, new ShortName(".", ""));
		dotEntry.setDirectory();
		dotEntry.setStartCluster(newStartCluster);
		FatLfnDirectoryEntry.copyDateTime(entry, dotEntry);
		result.addEntry(dotEntry, dotEntry.getActualEntry());
		
		// Second the dotdot entry which points to the parent directory (this)
		// if parent is the root dir then set start cluster to zero
		FatLfnDirectoryEntry dotDotEntry = FatLfnDirectoryEntry.createNew(null, new ShortName("..", ""));
		dotDotEntry.setDirectory();
		dotDotEntry.setStartCluster(isRoot() ? 0 : entry.getStartCluster());
		FatLfnDirectoryEntry.copyDateTime(entry, dotDotEntry);
		result.addEntry(dotDotEntry, dotDotEntry.getActualEntry());
		
		result.write();
		
		return result;
	}
	
	@Override
	public void setLength(long newLength) {
		throw new UnsupportedOperationException("This is a directory!");
	}
	
	@Override
	public long getLength() {
		throw new UnsupportedOperationException("This is a directory!");
	}
	
	@Override
	public boolean isDirectory() {
		return true;
	}

	@Override
	public String getName() {
		return entry.getName();
	}

	@Override
	public void setName(String newName) throws IOException {
		if(isRoot()) throw new IllegalStateException("Cannot rename root dir!");
		parent.renameEntry(entry, newName);
	}

	@Override
	public UsbFile getParent() {
		return parent;
	}

	@Override
	public String[] list() throws IOException {
		init();
		String[] list = new String[entries.size()];
		for(int i = 0; i < entries.size(); i++)
			list[i] = entries.get(i).getName();
		return list;
	}

	@Override
	public UsbFile[] listFiles() throws IOException {
		init();
		UsbFile[] list = new UsbFile[entries.size()];
		for(int i = 0; i < entries.size(); i++) {
			FatLfnDirectoryEntry entry = entries.get(i);
			if(entry.isDirectory()) {
				list[i] = FatDirectory.create(entry, blockDevice, fat, bootSector, this);
			} else {
				list[i] = FatFile.create(entry, blockDevice, fat, bootSector, this);
			}
		}
		return list;
	}

	@Override
	public void read(long offset, ByteBuffer destination) throws IOException {
		throw new UnsupportedOperationException("This is a directory!");
	}

	@Override
	public void write(long offset, ByteBuffer source) throws IOException {
		throw new UnsupportedOperationException("This is a directory!");
	}

	@Override
	public void flush() throws IOException {
		throw new UnsupportedOperationException("This is a directory!");
	}

	@Override
	public void close() throws IOException {
		throw new UnsupportedOperationException("This is a directory!");
	}

	@Override
	public void delete() throws IOException {
		if(isRoot()) throw new IllegalStateException("Root dir cannot be deleted!");
		
		init();
		parent.removeEntry(entry);
		parent.write();
		chain.setLength(0);
	}
}

File: libaums/src/main/java/com/github/mjdev/libaums/fs/fat32/FatDirectory.kt
/*
 * (C) Copyright 2014-2016 mjahnen <github@mgns.tech>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * 
 */

package com.github.mjdev.libaums.fs.fat32

import android.util.Log
import com.github.mjdev.libaums.driver.BlockDeviceDriver
import com.github.mjdev.libaums.fs.AbstractUsbFile
import com.github.mjdev.libaums.fs.UsbFile
import java.io.IOException
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.*


/**
 * This class represents a directory in the FAT32 file system. It can hold other
 * directories and files.
 *
 * @author mjahnen
 */
class FatDirectory
/**
 * Constructs a new FatDirectory with the given information.
 *
 * @param blockDevice
 * The block device the fs is located.
 * @param fat
 * The FAT of the fs.
 * @param bootSector
 * The boot sector if the fs.
 * @param parent
 * The parent directory of the newly created one.
 */
internal constructor(
            private val fs: Fat32FileSystem,
            private val blockDevice: BlockDeviceDriver,
            private val fat: FAT,
            private val bootSector: Fat32BootSector,
            /**
             * null if directory
             */
            private var entry: FatLfnDirectoryEntry?,
            /**
             * Null if this is the root directory.
             */
            override var parent: FatDirectory?) : AbstractUsbFile() {

    private lateinit var chain: ClusterChain
    /**
     * Entries read from the device.
     */
    private var entries: MutableList<FatLfnDirectoryEntry>? = null
    /**
     * Map for checking for existence when for example creating new files or
     * directories.
     *
     *
     * All items are stored in lower case because a FAT32 fs is not case
     * sensitive.
     */
    private val lfnMap: MutableMap<String, FatLfnDirectoryEntry>
    /**
     * Map for checking for existence of short names when generating short names
     * for new files or directories.
     */
    private val shortNameMap: MutableMap<ShortName, FatDirectoryEntry>

    /**
     * This method returns the volume label which can be stored in the root
     * directory of a FAT32 file system.
     *
     * @return The volume label.
     */
    internal var volumeLabel: String? = null
        private set

    private var hasBeenInited: Boolean = false

    /**
     *
     * @return True if this directory is the root directory.
     */
    override val isRoot: Boolean
        get() = entry == null

    override var length: Long
        get() = throw UnsupportedOperationException("This is a directory!")
        set(_) = throw UnsupportedOperationException("This is a directory!")

    override val isDirectory: Boolean
        get() = true

    override var name: String
        get() = if (entry != null) entry!!.name else "/"
        @Throws(IOException::class)
        set(newName) {
            check(!isRoot) { "Cannot rename root dir!" }
            parent!!.renameEntry(entry, newName)
        }

    init {
        lfnMap = HashMap()
        shortNameMap = HashMap()
    }

    /**
     * Initializes the [FatDirectory]. Creates the cluster chain if needed
     * and reads all entries from the cluster chain.
     *
     * @throws IOException
     * If reading from the device fails.
     */
    @Throws(IOException::class)
    private fun init() {
        if (!::chain.isInitialized) {
            chain = ClusterChain(entry!!.startCluster, blockDevice, fat, bootSector)
        }

        // entries is allocated here
        // an exception will be thrown if entries is used before the directory has been initialised
        // use of uninitialised entries can lead to data loss!
        if (entries == null) {
            entries = ArrayList()
        }

        // only read entries if we have no entries
        // otherwise newly created directories (. and ..) will read trash data
        if (entries!!.size == 0 && !hasBeenInited) {
            readEntries()
        }

        hasBeenInited = true
    }

    /**
     * Reads all entries from the directory and saves them into [.lfnMap],
     * [.entries] and [.shortNameMap].
     *
     * @throws IOException
     * If reading from the device fails.
     * @see .write
     */
    @Throws(IOException::class)
    private fun readEntries() {
        val buffer = ByteBuffer.allocate(chain.length.toInt())
        chain.read(0, buffer)
        // we have to buffer all long filename entries to parse them later
        val list = ArrayList<FatDirectoryEntry>()
        buffer.flip()
        while (buffer.remaining() > 0) {
            val e = FatDirectoryEntry.read(buffer) ?: break

            if (e.isLfnEntry) {
                list.add(e)
                continue
            }

            if (e.isVolumeLabel) {
                if (!isRoot) {
                    Log.w(TAG, "volume label in non root dir!")
                }
                volumeLabel = e.volumeLabel
                Log.d(TAG, "volume label: " + volumeLabel!!)
                continue
            }

            // we just skip deleted entries
            if (e.isDeleted) {
                list.clear()
                continue
            }

            val lfnEntry = FatLfnDirectoryEntry.read(e, list)
            addEntry(lfnEntry, e)
            list.clear()
        }
    }

    /**
     * Adds the long file name entry to [.lfnMap] and [.entries] and
     * the actual entry to [.shortNameMap].
     *
     *
     * This method does not write the changes to the disk. If you want to do so
     * call [.write] after adding an entry.
     *
     * @param lfnEntry
     * The long filename entry to add.
     * @param entry
     * The corresponding short name entry.
     * @see .removeEntry
     */
    private fun addEntry(lfnEntry: FatLfnDirectoryEntry, entry: FatDirectoryEntry) {
        entries!!.add(lfnEntry)
        lfnMap[lfnEntry.name.toLowerCase(Locale.getDefault())] = lfnEntry
        shortNameMap[entry.shortName!!] = entry
    }

    /**
     * Removes (if existing) the long file name entry from [.lfnMap] and
     * [.entries] and the actual entry from [.shortNameMap].
     *
     *
     * This method does not write the changes to the disk. If you want to do so
     * call [.write] after adding an entry.
     *
     * @param lfnEntry
     * The long filename entry to remove.
     * @see .addEntry
     */
    internal fun removeEntry(lfnEntry: FatLfnDirectoryEntry?) {
        entries!!.remove(lfnEntry)
        lfnMap.remove(lfnEntry!!.name.toLowerCase(Locale.getDefault()))
        shortNameMap.remove(lfnEntry.actualEntry.shortName)
    }

    /**
     * Renames a long filename entry to the desired new name.
     *
     *
     * This method immediately writes the change to the disk, thus no further
     * call to [.write] is needed.
     *
     * @param lfnEntry
     * The long filename entry to rename.
     * @param newName
     * The new name.
     * @throws IOException
     * If writing the change to the disk fails.
     */
    @Throws(IOException::class)
    internal fun renameEntry(lfnEntry: FatLfnDirectoryEntry?, newName: String) {
        if (lfnEntry!!.name == newName)
            return

        removeEntry(lfnEntry)
        lfnEntry.setName(newName,
                ShortNameGenerator.generateShortName(newName, shortNameMap.keys))
        addEntry(lfnEntry, lfnEntry.actualEntry)
        write()
    }

    /**
     * Writes the [.entries] to the disk. Any changes made by
     * [.addEntry] or
     * [.removeEntry] will then be committed to the
     * device.
     *
     * @throws IOException
     * @see {@link .write
     */
    @Throws(IOException::class)
    internal fun write() {
        init()
        val writeVolumeLabel = isRoot && volumeLabel != null
        // first lookup the total entries needed
        var totalEntryCount = 0
        for (entry in entries!!) {
            totalEntryCount += entry.entryCount
        }

        if (writeVolumeLabel)
            totalEntryCount++

        val totalBytes = (totalEntryCount * FatDirectoryEntry.SIZE).toLong()
        chain.length = totalBytes

        val buffer = ByteBuffer.allocate(chain.length.toInt())
        buffer.order(ByteOrder.LITTLE_ENDIAN)

        if (writeVolumeLabel)
            FatDirectoryEntry.createVolumeLabel(volumeLabel!!).serialize(buffer)

        for (entry in entries!!) {
            entry.serialize(buffer)
        }

        if (totalBytes % bootSector.bytesPerCluster != 0L || totalBytes == 0L) {
            // add dummy entry filled with zeros to mark end of entries
            buffer.put(ByteArray(buffer.remaining()))
        }

        buffer.flip()
        chain.write(0, buffer)
    }

    @Throws(IOException::class)
    override fun createFile(name: String): FatFile {
        if (lfnMap.containsKey(name.toLowerCase(Locale.getDefault())))
            throw IOException("Item already exists!")

        init() // initialise the directory before creating files

        val shortName = ShortNameGenerator.generateShortName(name, shortNameMap.keys)

        val entry = FatLfnDirectoryEntry(name, shortName)
        // alloc completely new chain
        val newStartCluster = fat.alloc(arrayOf(), 1)[0]
        entry.startCluster = newStartCluster

        Log.d(TAG, "adding entry: $entry with short name: $shortName")
        addEntry(entry, entry.actualEntry)
        // write changes immediately to disk
        write()

        val file = FatFile(blockDevice, fat, bootSector, entry, this)
        fs.fileCache[file.absolutePath] = file
        return file
    }

    @Throws(IOException::class)
    override fun createDirectory(name: String): FatDirectory {
        if (lfnMap.containsKey(name.toLowerCase(Locale.getDefault())))
            throw IOException("Item already exists!")

        init() // initialise the directory before creating files

        val shortName = ShortNameGenerator.generateShortName(name, shortNameMap.keys)

        val entry = FatLfnDirectoryEntry(name, shortName)
        entry.setDirectory()
        // alloc completely new chain
        val newStartCluster = fat.alloc(arrayOf(), 1)[0]
        entry.startCluster = newStartCluster

        Log.d(TAG, "adding entry: $entry with short name: $shortName")
        addEntry(entry, entry.actualEntry)
        // write changes immediately to disk
        write()

        val result = FatDirectory(fs, blockDevice, fat, bootSector, entry, this)
        result.hasBeenInited = true

        result.entries = ArrayList() // initialise entries before adding sub-directories

        // first create the dot entry which points to the dir just created
        val dotEntry = FatLfnDirectoryEntry(null, ShortName(".", ""))
        dotEntry.setDirectory()
        dotEntry.startCluster = newStartCluster
        FatLfnDirectoryEntry.copyDateTime(entry, dotEntry)
        result.addEntry(dotEntry, dotEntry.actualEntry)

        // Second the dotdot entry which points to the parent directory (this)
        // if parent is the root dir then set start cluster to zero
        val dotDotEntry = FatLfnDirectoryEntry(null, ShortName("..",
                ""))
        dotDotEntry.setDirectory()
        dotDotEntry.startCluster = if (isRoot) 0 else this.entry!!.startCluster
        FatLfnDirectoryEntry.copyDateTime(entry, dotDotEntry)
        result.addEntry(dotDotEntry, dotDotEntry.actualEntry)

        // write changes immediately to disk
        result.write()

        fs.fileCache[result.absolutePath] = result
        return result
    }

    override fun createdAt(): Long {
        check(!isRoot) { "root dir!" }
        return entry!!.actualEntry.createdDateTime
    }

    override fun lastModified(): Long {
        check(!isRoot) { "root dir!" }
        return entry!!.actualEntry.lastModifiedDateTime
    }

    override fun lastAccessed(): Long {
        check(!isRoot) { "root dir!" }
        return entry!!.actualEntry.lastAccessedDateTime
    }

    @Throws(IOException::class)
    override fun list(): Array<String> {
        init()
        val list = ArrayList<String>(entries!!.size)
        for (entry in entries!!) {
            val name = entry.name
            if (name != "." && name != "..") {
                list.add(name)
            }
        }

        return list.toTypedArray()
    }

    @Throws(IOException::class)
    override fun listFiles(): Array<UsbFile> {
        init()
        val list = ArrayList<UsbFile>(entries!!.size)
        for (entry in entries!!) {
            val name = entry.name
            if (name == "." || name == "..")
                continue

            val entryAbsolutePath = if (isRoot) {
                UsbFile.separator + entry.name
            } else {
                absolutePath + UsbFile.separator + entry.name
            }

            val file = when {
                fs.fileCache[entryAbsolutePath] != null -> fs.fileCache[entryAbsolutePath]!!
                entry.isDirectory -> FatDirectory(fs, blockDevice, fat, bootSector, entry, this)
                else -> FatFile(blockDevice, fat, bootSector, entry, this)
            }

            fs.fileCache[entryAbsolutePath] = file
            list.add(file)
        }

        return list.toTypedArray()
    }

    @Throws(IOException::class)
    override fun read(offset: Long, destination: ByteBuffer) {
        throw UnsupportedOperationException("This is a directory!")
    }

    @Throws(IOException::class)
    override fun write(offset: Long, source: ByteBuffer) {
        throw UnsupportedOperationException("This is a directory!")
    }

    @Throws(IOException::class)
    override fun flush() {
        throw UnsupportedOperationException("This is a directory!")
    }

    @Throws(IOException::class)
    override fun close() {
        throw UnsupportedOperationException("This is a directory!")
    }

    @Throws(IOException::class)
    override fun moveTo(destination: UsbFile) {
        check(!isRoot) { "cannot move root dir!" }
        check(destination.isDirectory) { "destination cannot be a file!" }
        check(destination is FatDirectory) { "cannot move between different filesystems!" }

        // TODO check if destination is really on the same physical device or
        // partition!

        if (destination.lfnMap.containsKey(entry!!.name.toLowerCase(Locale.getDefault())))
            throw IOException("item already exists in destination!")

        init()
        destination.init()

        // now the actual magic happens!
        parent!!.removeEntry(entry)
        destination.addEntry(entry!!, entry!!.actualEntry)

        parent!!.write()
        destination.write()
        parent = destination
    }

    /**
     * This method moves an long filename entry currently stored in THIS
     * directory to the destination which also must be a directory.
     *
     *
     * Used by [FatFile] to move itself to another directory.
     *
     * @param entry
     * The entry which shall be moved.
     * @param destination
     * The destination directory.
     * @throws IOException
     * If writing fails or the item already exists in the
     * destination directory.
     * @throws IllegalStateException
     * If the destination is not a directory or destination is on a
     * different file system.
     */
    @Throws(IOException::class)
    internal fun move(entry: FatLfnDirectoryEntry, destination: UsbFile) {
        check(destination.isDirectory) { "destination cannot be a file!" }
        check(destination is FatDirectory) { "cannot move between different filesystems!" }

        // TODO check if destination is really on the same physical device or
        // partition!

        if (destination.lfnMap.containsKey(entry.name.toLowerCase(Locale.getDefault())))
            throw IOException("item already exists in destination!")

        init()
        destination.init()

        // now the actual magic happens!
        removeEntry(entry)
        destination.addEntry(entry, entry.actualEntry)

        write()
        destination.write()
    }

    @Throws(IOException::class)
    override fun delete() {
        check(!isRoot) { "Root dir cannot be deleted!" }

        init()
        val subElements = listFiles()

        for (file in subElements) {
            file.delete()
        }

        parent!!.removeEntry(entry)
        parent!!.write()
        chain.length = 0
    }

    companion object {

        private val TAG = FatDirectory::class.java.simpleName

        /**
         * Reads the root directory from a FAT32 file system.
         *
         * @param blockDevice
         * The block device the fs is located.
         * @param fat
         * The FAT of the fs.
         * @param bootSector
         * The boot sector if the fs.
         * @return Newly created root directory.
         * @throws IOException
         * If reading from the device fails.
         */
        @JvmStatic
        @Throws(IOException::class)
        internal fun readRoot(fs: Fat32FileSystem,
                              blockDevice: BlockDeviceDriver, fat: FAT,
                              bootSector: Fat32BootSector): FatDirectory {
            val result = FatDirectory(fs, blockDevice, fat, bootSector, null, null)
            result.chain = ClusterChain(bootSector.rootDirStartCluster, blockDevice, fat,
                    bootSector)
            result.init() // init calls readEntries
            return result
        }
    }
}
